advanced_topics.qmd

```diff
@@ -266,15 +277,359 @@ pums_person |>
   collect()
 ```
 
+### Using UDFs for regression model predictions
+
+Another example where UDFs can be useful is where you've created a model based on a subset of your data and you want to use it to make predictions.
+Let's say we've created a linear model, using education and employment status to predict hours worked per week for respondents in 2018.
+
+```{r}
+#| label: model-commute-time
+model_data <- open_dataset("data/person/") |>
+    filter(year == 2018) |>
+    filter(AGEP >= 16, AGEP < 65, !is.na(COW)) |>
+    transmute(
+      self_employed = str_detect(as.character(COW), "Self-employed"),
+      edu = case_when(
+        grepl("bachelor|associate|professional|doctorate|master", SCHL, TRUE) ~
+          "College or higher",
+        grepl("college", SCHL, TRUE) ~
+          "Some college",
+        grepl("High school graduate|high school diploma|GED", SCHL, TRUE) ~
+          "High school",
+        TRUE ~ "Less than High school"
+      ),
+      hours_worked = WKHP,
+      PWGTP = PWGTP
+    ) |>
+    collect()
+
+hours_worked_model <- survey::svyglm(
+  hours_worked ~ self_employed * edu,
+  design = survey::svydesign(ids = ~1, weights = ~PWGTP, data = model_data)
+)
+summary(hours_worked_model)
+```
+
+We can use this model to predict what our model would have on this data.
+We use the `predict` function to make these predictions and we also calculate the weighted mean for each group.
+This is going to be very very close to the actual weighted means since this data was used to fit this linear model.
+
+```{r}
+#| label: summarise_fit_data
+#| warning: false
+model_data$predicted <- predict(hours_worked_model, newdata = model_data)
+
+model_data_summary <- model_data |>
+  group_by(edu, self_employed) |>
+  summarise(
+    work_hours_mean = weighted.mean(hours_worked, w = PWGTP, na.rm = TRUE),
+    work_hours_predicted = weighted.mean(predicted, w = PWGTP),
+  )
+```
+
+And we can visualize the underlying data along with weighted mean and predictions.
```

Comment:
This is me being 👴 I think but I hate this use of the word "visualize". Visualizing to me means closing your eyes and imagining something.

Reply:
Heh, happy for any other verb